src/python/module.cpp

```diff
@@ -1,10 +1,765 @@
+      .def("xx_plus_yy",
+           nb::overload_cast<qc::Qubit, qc::Qubit, qc::fp, qc::fp>(
+               &qc::QuantumComputation::xx_plus_yy))
+      .def("xx_plus_yy",
+           nb::overload_cast<qc::Qubit, qc::Qubit, const qc::Control&, qc::fp,
+                             qc::fp>(&qc::QuantumComputation::xx_plus_yy))
+      .def("xx_plus_yy",
+           nb::overload_cast<qc::Qubit, qc::Qubit, const qc::Controls&, qc::fp,
+                             qc::fp>(&qc::QuantumComputation::xx_plus_yy))
+      .def("measure", nb::overload_cast<qc::Qubit, std::size_t>(
+                          &qc::QuantumComputation::measure))
+      .def("measure",
+           nb::overload_cast<qc::Qubit, const std::pair<std::string, qc::Bit>&>(
+               &qc::QuantumComputation::measure))
+      .def("measure", nb::overload_cast<const std::vector<qc::Qubit>&,
+                                        const std::vector<qc::Bit>&>(
+                          &qc::QuantumComputation::measure))
+      .def("reset",
+           nb::overload_cast<qc::Qubit>(&qc::QuantumComputation::reset))
+      .def("reset", nb::overload_cast<const std::vector<qc::Qubit>&>(
+                        &qc::QuantumComputation::reset))
+      .def("barrier",
+           nb::overload_cast<qc::Qubit>(&qc::QuantumComputation::barrier))
+      .def("barrier", nb::overload_cast<const std::vector<qc::Qubit>&>(
+                          &qc::QuantumComputation::barrier))
+      .def("classic_controlled",
+           nb::overload_cast<const qc::OpType, const qc::Qubit,
+                             const qc::ClassicalRegister&, const std::uint64_t,
+                             const std::vector<qc::fp>&>(
+               &qc::QuantumComputation::classicControlled))
+      .def("classic_controlled",
+           nb::overload_cast<const qc::OpType, const qc::Qubit,
+                             const qc::Control, const qc::ClassicalRegister&,
+                             const std::uint64_t, const std::vector<qc::fp>&>(
+               &qc::QuantumComputation::classicControlled))
+      .def("classic_controlled",
+           nb::overload_cast<const qc::OpType, const qc::Qubit,
+                             const qc::Controls&, const qc::ClassicalRegister&,
+                             const std::uint64_t, const std::vector<qc::fp>&>(
+               &qc::QuantumComputation::classicControlled))
+      .def("set_logical_qubit_ancillary",
+           &qc::QuantumComputation::setLogicalQubitAncillary)
+      .def("add_qubit_register", &qc::QuantumComputation::addQubitRegister)
+      .def("add_classical_bit_register",
+           &qc::QuantumComputation::addClassicalRegister)
+      .def("add_ancillary_register",
+           &qc::QuantumComputation::addAncillaryRegister)
+      .def("append_operation",
+           [](QuantumComputation& qc, const qc::Operation& op) {
+             qc.emplace_back(op.clone()); // not an ideal solution but it works
```

Comment:
No matter what I tried, this doesn't work.

Moving anything but a `unique_ptr` to `Operation` doesn't even compile

Moving a `unique_ptr` compiles but crashes when invoking the method.

Reply:
Hm. Let's keep this thread open for now. There has to be some way to accomplish that. I probably won't have much time to dig into that myself the coming days but would love to at some point.

Reply:
pybind doesn't relinquish ownership of pointers so we cannot move. One could pass a raw pointer and create a unique_ptr from that. But that would break the intention of the unique_ptr and python would still delete the raw pointer anyway. Unless the *entire* implementation is switched to shared_ptr there is no way around the copy here.

Reply:
Doesn't sound too promising 😢  i'll still give it a look during my review.